File: workspaces/fp-0002-shpigovsky-v9/tools/o-centre-visual-discrepancy-audit/micro-pass-01-qa.py
"""Technical QA for O-Centre micro-pass 01."""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

from playwright.sync_api import sync_playwright

V8 = Path(r"X:\AI MARS\workspaces\fp-0002-shpigovsky-v8")
DIST = V8 / "dist"
PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 4722
URL = f"http://127.0.0.1:{PORT}/o-centre.html"


def main() -> None:
    html = (DIST / "o-centre.html").read_text(encoding="utf-8")
    ids = re.findall(r'\bid="([^"]+)"', html)
    dup_ids = sorted({i for i in ids if ids.count(i) > 1})
    includes_unresolved = html.count("@@include")
    asset_refs = re.findall(r'(?:src|href)="(assets/[^"]+)"', html)
    missing_assets = [a for a in asset_refs if not (DIST / a).exists()]

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page(viewport={"width": 1437, "height": 900})
        errors: list[str] = []
        page.on("pageerror", lambda e: errors.append(str(e)))
        page.on("console", lambda msg: errors.append(msg.text) if msg.type == "error" else None)
        page.goto(URL, wait_until="load", timeout=60000)
        page.wait_for_timeout(3000)
        overflow = page.evaluate(
            "() => document.documentElement.scrollWidth > window.innerWidth"
        )
        browser.close()

    out = {
        "build_dist_exists": DIST.exists(),
        "unresolved_includes": includes_unresolved,
        "duplicate_ids": dup_ids,
        "missing_assets": missing_assets,
        "console_errors": errors,
        "horizontal_overflow_1437": overflow,
    }
    print(json.dumps(out, ensure_ascii=False, indent=2))
    if includes_unresolved or missing_assets or dup_ids or errors or overflow:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: workspaces/fp-0002-shpigovsky-v9/tools/o-centre-visual-discrepancy-audit/micro-pass-01-qa.mjs
import { readFileSync } from "fs";
import { join, resolve, dirname } from "path";
import { fileURLToPath } from "url";
import puppeteer from "puppeteer-core";

const __dir = dirname(fileURLToPath(import.meta.url));
const v8Root = resolve(__dir, "../..");
const distDir = join(v8Root, "dist");
const PORT = 4722;
const CHROME = "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe";

const html = readFileSync(join(distDir, "o-centre.html"), "utf-8");
const ids = [...html.matchAll(/\bid="([^"]+)"/g)].map((m) => m[1]);
const dupIds = [...new Set(ids.filter((id, i) => ids.indexOf(id) !== i))].sort();
const assets = [...html.matchAll(/(?:src|href)="(assets\/[^"]+)"/g)].map((m) => m[1]);
const { existsSync } = await import("fs");
const missingAssets = assets.filter((a) => !existsSync(join(distDir, a)));

import { createServer } from "http";
const server = await new Promise((resolvePromise) => {
  const s = createServer((req, res) => {
    let rel = decodeURIComponent(req.url.split("?")[0]);
    if (rel === "/") rel = "/o-centre.html";
    const filePath = join(distDir, rel.replace(/^\//, ""));
    try {
      const data = readFileSync(filePath);
      const ext = filePath.split(".").pop();
      const types = { html: "text/html; charset=utf-8", css: "text/css", js: "application/javascript", webp: "image/webp" };
      res.writeHead(200, { "Content-Type": types[ext] || "application/octet-stream" });
      res.end(data);
    } catch {
      res.writeHead(404);
      res.end("not found");
    }
  });
  s.listen(PORT, "127.0.0.1", () => resolvePromise(s));
});

const browser = await puppeteer.launch({ executablePath: CHROME, headless: true, args: ["--no-sandbox"] });
const page = await browser.newPage();
const consoleErrors = [];
page.on("pageerror", (e) => consoleErrors.push(String(e)));
page.on("console", (msg) => { if (msg.type() === "error") consoleErrors.push(msg.text()); });
await page.setViewport({ width: 1437, height: 900 });
await page.goto(`http://127.0.0.1:${PORT}/o-centre.html`, { waitUntil: "networkidle0", timeout: 60000 });
const overflow = await page.evaluate(() => document.documentElement.scrollWidth > window.innerWidth);
await browser.close();
server.close();

console.log(JSON.stringify({
  build: "PASS",
  unresolved_includes: (html.match(/@@include/g) || []).length,
  duplicate_ids: dupIds,
  missing_assets: missingAssets,
  console_errors: consoleErrors,
  horizontal_overflow_1437: overflow,
}, null, 2));
